mlir/include/mlir/Dialect/XeGPU/uArch/uArch.h

```diff
@@ -0,0 +1,389 @@
+#include <map>
+#include <mutex>
+#include <shared_mutex>
+#include <tuple>
+
+#include "mlir/IR/Types.h"
+
+namespace mlir {
+namespace xegpu {
+namespace uArch {
+
+// Data types we need for YAML to uArch translation
+struct Range {
+  int start;
+  int end;
+};
+
+// Restriction struct
+// This struct is used to represent a restriction on the uArch
+// The restriction is represented as a range of necessary parameters (template
+// arguments) and a lambda function (validate()) that takes the same number of
+// arguments as the number of template arguments The lambda function returns
+// true if the arguments satisfy the restriction The lambda function returns
+// false if the arguments do not satisfy the restriction
+
+// For example, a restriction that checks if the number of dimensions in a
+// std::vector<std::vector<uint32_t>> is 2 can be represented as:
+// std::vector<std::vector<uint32_t>> rt =
+// {{1, 32}, {2, 16}}; Restriction<std::vector<std::vector<uint32_t>>> r1(rt,
+// [](std::vector<std::vector<uint32_t>> t) { return t.size() == 2; });
+// r1.validate() will return true if the number of dimensions in the
+// std::vector<std::vector<uint32_t>> is 2 r1.validate() will return false if
+// the number of dimensions in the std::vector<std::vector<uint32_t>> is not 2
+
+// The primary purpose of Restriction struct is to provide a generic way to
+// represent restrictions on the uArch and to validate if the uArch satisfies
+// the restrictions
+template <typename... Args>
+struct Restriction {
+  std::tuple<Args...> data;
+  std::function<bool(Args...)> func;
+
+  Restriction(Args... args, std::function<bool(Args...)> f)
+      : data(args...), func(f) {}
+
+  bool validate() { return std::apply(func, data); }
+  std::any apply() { return std::apply(func, data); }
+};
+
+// Architecture HW component hierarchy to present thread, core, socket ...
+struct uArchHierarchyComponent {
+  std::string name = ""; // optional name of the hierarchy component
+  // no. of lower hierarchy component it contains, e.g., for PVC XeCore it
+  // contains 8 threads, so no_of_component=8
+  uint32_t no_of_component;
+  // Constructor
+  uArchHierarchyComponent(const std::string &name, uint32_t no_of_component)
+      : name(name), no_of_component(no_of_component) {}
+};
+
+// An enum class to represent the functional unit of an instruction
+enum class FunctionalUnit {
+  ALU,
+  Tensor,
+  Matrix,
+  Load,
+  Store,
+  Branch,
+  Barrier,
+  Memory,
+  Atomic,
+  Interconnect,
+  Other
+};
+
+// An enum class to represent the type of memory
+enum class MemoryType { Shared, Local, Global, Constant, Texture, Other };
+
+// An enum class to represent the memory access type
+enum class MemoryAccessType { Read, Write, ReadWrite, Other };
+
+// An enum class to represent the type of an instruction
+enum class InstructionType { SIMT, SIMD, SPMD, MIMD, Other };
+
+// An enum class to represent the scope of an instruction
+enum class InstructionScope {
```

Comment:
What is the reason the instruction itself doesn't indicate the scope?

Reply:
For user to know at what scope the instruction is native to (work-item, subgroup or workgroup)